packages/node_modules/function-tree/README.md

```diff
@@ -341,6 +341,121 @@ const ft = new FunctionTree([
 
 If the condition returns true the execution will throw an error and stop.
 
+### tags
+Tags gives you more power to write declarative code. You combine them with function factories to point to values at different targets. The template tags can compose other tags.
+
+#### ResolveProvider
+By default your function tree instance includes the ResolveProvider. It allows you to evaluate tags.
+
+```js
+function myFunctionFactory(someTag) {
+  function myFunction ({resolve}) {
+    // Evaluates to the tags value, if not a tag it will just return the value
+    resolve.value(someTag)
+
+    // Evaluates the path in the tag
+    resolve.path(someTag)
+
+    // Evaluates if variable is a tag, can optionally pass array of types
+    // ['props', 'string'] as second argument
+    resolve.isTag(someTag)
+
+    // Evaluates if variable is a TagValue. Useful when you have created custom
+    // stuff to be evaluated in tags, extending TagValue class
+    resolve.isTagValue(someTag)
+  }
+
+  return myFunction
+}
+```
+
+#### props
+Targets the props passed into the execution of a function tree:
+
+```js
+import {props} from 'function-tree/tags'
+
+export default [
+  shout(props`text`)
+]
+```
+
+#### string
+Returns the evaluated string.
+
+```js
+import {string, props} from 'function-tree/tags'
+
+export default [
+  shout(string`Hello ${props`name`}`)
+]
+```
+
+#### createTemplateTag
+Allows you to add new tags, targeting your custom providers.
+
+```js
+import {createTemplateTag} from 'function-tree/tags'
+
+export const myTag = createTemplateTag('myTag', (path, context) => {
+  return context.myProvider.get(path)
+})
+```
+
+#### convertObjectWithTemplates
+Converts an object passed to a function factory, resolving any template tags.
+
+```js
+import {convertObjectWithTemplates} from 'function-tree/tags'
+
+function myFunctionFactory(someObject) {
+  function myFunction({resolve}) {
+    const object = convertObjectWithTemplates(someObject, resolve)
+  }
+}
+```
+
+#### extractValueWithPath
+Easy way to support nested values extraction from objects using the path.
+
+```js
+import {createTemplateTag, extractValueWithPath} from 'function-tree/tags'
+
+export const myTag = createTemplateTag('myTag', (path, context) => {
+  return extractValueWithPath(context.something, path)
+})
+```
+
+#### TagValue
+Allows you to create a custom tag value container.
+
+```js
+import {TagValue} from 'function-tree/tags'
+
+class Uppercaser extends TagValue {
+  constructor(tag) {
+    super()
+    this.tag = tag
+  }
+  // Required method
+  getValue(context) {
+    return this.tag.getValue(context).toUpperCase()
+  }
+}
+
+export default (tag) => {
+  return new Uppercaser(tag)
+}
+```
+
+```js
+import uppercaser from './uppercaser'
+
+export default [
+  shout(string`Hello, ${uppercaser(props`name`)}`)
```

Comment:
Not sure I understand here. Is `uppercaser` a tag now ? Or is it just a wrapper around another tag ?

Reply:
How. OK. TagValue is just a wrapper around `getValue`. I tried to implement such dynamic values by creating custom tags but it was a mess. Very cool idea to make this more accessible.

But I think the name `TagValue` is misleading: it has nothing to do with tagged templates. I would call this `DynamicValue` because this is what it gives us: a value that is resolved when signal is executed.

Here are two examples that I stumbled on and for which I wanted such a thing:

```js
class Now extends DynamicValue {
  getValue() {
    return Date.now()
  }
}
export const now = new Now()

class Uuid extends DynamicValue {
  getValue() {
    return uuid.v4()
  }
}
export const uuid = new Uuid()

const mychain = [
  set(props`now`, now),
  set(props`uid`, uuid)
]
```

Also I think there is no need to call `super` in examples...